snuba/migrations/snuba_migrations/sessions/0002_sessions_aggregates.py

```diff
@@ -0,0 +1,281 @@
+from typing import Sequence
+
+from snuba.clickhouse.columns import (
+    AggregateFunction,
+    Column,
+    DateTime,
+    String,
+    UInt,
+    UUID,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.processor import MAX_UINT32, NIL_UUID
+
+DEFAULT_QUANTITY = 1
+
+
+v1_matview: operations.Operation = import_module(
+    "snuba.migrations.snuba_migrations.sessions.0001_sessions"
+).create_matview
+
+
+aggregate_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("started", DateTime()),
+    Column("release", String(Modifiers(low_cardinality=True))),
+    Column("environment", String(Modifiers(low_cardinality=True))),
+    Column(
+        "duration_quantiles",
+        AggregateFunction("quantilesIf(0.5, 0.9)", [UInt(32), UInt(8)]),
+    ),
+    Column("duration_avg", AggregateFunction("avgIf", [UInt(32), UInt(8)])),
+    # sessions:
+    Column("sessions", AggregateFunction("countIf", [UUID(), UInt(8)])),
+    Column("sessions_sum", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    Column("sessions_crashed", AggregateFunction("countIf", [UUID(), UInt(8)])),
+    Column("sessions_crashed_sum", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    Column("sessions_abnormal", AggregateFunction("countIf", [UUID(), UInt(8)])),
+    Column("sessions_abnormal_sum", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    Column("sessions_errored", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("sessions_errored_sum", AggregateFunction("sumIf", [UInt(32), UInt(8)])),
+    # users:
+    Column("users", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("users_crashed", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("users_abnormal", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+    Column("users_errored", AggregateFunction("uniqIf", [UUID(), UInt(8)])),
+]
+
+
+query = f"""
+SELECT
+  org_id,
+  project_id,
+  toStartOfHour(started) as started,
+  release,
+  environment,
+
+  -- we would have liked to changed the quantiles here to `0.5, 0.75, 0.95, 0.99`:
+  quantilesIfState(0.5, 0.9)(
+    duration,
+    duration <> {MAX_UINT32} AND status == 1
+  ) as duration_quantiles,
+
+  -- this is new:
+  avgIfState(duration, duration <> {MAX_UINT32} AND status == 1) as duration_avg,
+
+  -- keep the old `count` as they were, just change their condition
+  countIfState(session_id, seq == 0 AND session_id != '{NIL_UUID}') as sessions,
+  countIfState(session_id, status == 2 AND session_id != '{NIL_UUID}') as sessions_crashed,
+  countIfState(session_id, status == 3 AND session_id != '{NIL_UUID}') as sessions_abnormal,
+  -- "old" individual sessions use the uniq session_id as before:
+  uniqIfState(session_id, errors > 0 AND session_id != '{NIL_UUID}') as sessions_errored,
```

Comment:
Actually I found an alternative approach that would allow us to eventually get rid of the old columns here.
Did you ever try to add the new columns to the destination table without changing the materialized view and consume some data through the consumer? 
Does that work ? If we were able to create those new columns empty we could go for this sequence:
1) add new columns in dest and source
2) change the query path to sum columns
3) noew rebuild the matview, when the new matview is done only the new columns will be filled and the counts of the old columns will always be 0. Of course the condition has to be different.

Eventually we can drop the old columns.

It has the downside that reverting the migration would be impossible without some data loss.
So I think we should stick with this at this point, but it would be important to understand if this path is possible at all.

Reply:
Can you clarify what you mean by "rebuild the matview" here? Anyhow, I think the non-destructive approach I’m using here is the way forward ;-)

Reply:
Hi, so discussed with @JTCunning, we agreed we can do the approach I mentioned above:
Since adding columns to the destination table without changing the matview now works, we should arrange this migration in the following step (each step a PR at least - if you want to split them more, that's fine)
1) Add the new columns to the source and destination tables. No change to the matview
2) Make the change to the query path, so now the query path merges the columns in the destination table and sums them with the old ones when needed. Of course the new columns are all empty at this point 
3) Rebuild the matview so it switches from filling the old columns to filling ONLY the new columns if the quantity is populated. (of course default cannot be 1 otherwise you cannot distinguish the individual session before the migration from the individual sessions after migration).  Now all sessions ingested are still individual and are only filling the old columns 
4) Deploy the new consumer that populates the quantity field, now the matview sees that and starts filling the new columns.

How does it sound ?

Reply:
whats the reason making the switchover based on the consumer in step 4, and not already in step 3? You mentioned the ability to roll back step 3 without losing data?

Reply:
This plan works if the quantity default is NULL. So that means we fill in the legacy columns with legacy data (coming from  the old consumer). This ensures that we are filling in the old table until we do the switch over. 
The consumer is the one that starts filling in the quantity: `1` for the new individual sessions, `anything else` for aggregated sessions. If we start consuming (and filling in the quantity) before we change the matview, the existing matview will still fill the legacy columns even if quantity is not NULL. Which will make the dest table inconsistent with the source. Not necessarily terrible, but undesirable. Thus 3 should be before 4.
Now they just cannot be done together since we cannot, in production, synchronize a deployment (new consumer) with a data migration (recreate the table) so one has to be done before the other and in such a way that when the first step is done the system still works.

We cannot roll back the matview change after step 4 because we will nto be filling in the legacy columns anymore, only the new ones.